master/pkg/searcher/asha.go

```diff
@@ -174,9 +175,13 @@ func (s *asyncHalvingSearch) trialClosed(
 }
 
 func (s *asyncHalvingSearch) validationCompleted(
-	ctx context, requestID model.RequestID, metric float64, op ValidateAfter,
+	ctx context, requestID model.RequestID, metrics interface{}, op ValidateAfter,
 ) ([]Operation, error) {
 	s.PendingTrials--
+	metric, ok := metrics.(float64)
+	if !ok {
+		return nil, fmt.Errorf("unexpected metric type for ASHA built-in search method %v", metrics)
```

Comment:
Sorry, I realized I had left this comment in the Pending stage so you couldn't see it (hit "start a review" instead of "add single comment")

Reply:
@rb-determined-ai wanted to ask for your opinion about this. In theory, maybe we could fish the name of the scalar metric out of the experiment config on the master. But, as you repeatedly said, the experiment config ain't a stable API.

Reply:
I have a "general-direction" sort of answer here, but I can't pretend to have a "crystal clear vision" sort of answer.

I really want our searcher to be dumber, and to be less integrated with the experiment config.  I would like for us to support a world where you might have some 3rd party searcher provider, and I don't want it to have to know about the experiment config and the `searcher.metric` there.

Additionally, I want our searcher config to not be the source of truth for "default sorting behavior".  Default sorting behavior is used by things like automatic checkpoint gc or "give me the top N checkpoints", which are features that should be available with any task that can report metrics and save checkpoints.  If we had generic tasks today, we'd have an issue where those tasks don't have a `searcher` config field and we wouldn't be able to provide that information.

Reply:
I'm not sure that the answer here has a bearing on this wider general-direction question?

The motivation is just for our currently-implemented searchers to expand the types of values they can process.  I would expect `float`s and `dict`s to be the most common `report_completed` types, so it would be really nice for all of our methods to accept both.  

While our current searchers would need to access `searcher.metric` in the config to know how to process a `dict`, I imagine that we could move to dumber, python searcher objects in the future and that these could just have the `metric` field in their constructor without any need to introspect some external config.  Maybe the choice could limit that future world, but I'm not seeing how.

Reply:
It just places more requirements on the searcher provider, and ties the searcher provider closer to the experiment config.

I'd really rather we not support that for now.  If in the future we have a better plan for how hpsearch works in a post-detached training world, and it's clearly a useful feature, we can add it later with no cost.  But if we add it now removing it will be a real pain.